spec/terminal-tab-spec.js

```diff
@@ -51,4 +59,92 @@ describe('TerminalTab', () => {
     });
 
   });
+
+  describe('when the terminal:focus event is triggered', () => {
+    it ('navigates to an open terminal', () => {
+      terminalOpenAndWait(workspaceElement);
+      let bottomDock = atom.workspace.getBottomDock();
+
+
+      runs(() => {
+        bottomDock.hide();
+        let terminalPromise = atom.commands.dispatch(workspaceElement, 'terminal:focus');
+        waitsForPromise(() => {
+          return terminalPromise;
+        });
+
+        // ensure the bottom dock is visible, and is the active item.
+        expect(bottomDock.isVisible()).toBe(true);
+
+        let activePaneItem = bottomDock.getActivePaneItem();
+        expect(activePaneItem).toBeInstanceOf(TerminalSession);
+
+        activePaneItem = atom.workspace.getActivePaneItem();
+        expect(activePaneItem).toBeInstanceOf(TerminalSession);
+      });
+    });
+
+    it ('opens and focuses a new terminal, if one does not exist', () => {
+      let bottomDock = atom.workspace.getBottomDock();
+      let terminalPromise = atom.commands.dispatch(workspaceElement, 'terminal:focus');
+      waitsForPromise(() => {
+        return terminalPromise;
+      });
+
+      runs(() => {
+        // ensure the bottom dock is visible, and is the active item.
+        expect(bottomDock.isVisible()).toBe(true);
+
+        let activePaneItem = bottomDock.getActivePaneItem();
+        expect(activePaneItem).toBeInstanceOf(TerminalSession);
+
+        activePaneItem = atom.workspace.getActivePaneItem();
+        expect(activePaneItem).toBeInstanceOf(TerminalSession);
+      });
+    });
+
+    it ('keeps focus if the terminal is already focused', () => {
+      terminalOpenAndWait(workspaceElement);
+
+      runs(() => {
+        activePaneItem = atom.workspace.getActivePaneItem();
+        expect(activePaneItem).toBeInstanceOf(TerminalSession);
+      });
+
+      let terminalPromise = atom.commands.dispatch(workspaceElement, 'terminal:focus');
+      waitsForPromise(() => {
+        return terminalPromise;
+      });
+
+      runs(() => {
+        activePaneItem = atom.workspace.getActivePaneItem();
+        expect(activePaneItem).toBeInstanceOf(TerminalSession);
+      });
+
+    });
+  });
+
+  describe('when the terminal:unfocus event is triggered', () => {
+    it ('returns focus to the previously active panel', () => {
+        let originalActivePane = atom.workspace.getActivePane();
+        let terminalView;
+        terminalOpenAndWait(workspaceElement);
+
+        runs(() => {
+          expect(atom.workspace.getActivePane()).toNotBe(originalActivePane);
+          let activePaneItem = atom.workspace.getActivePaneItem();
+          expect(activePaneItem).toBeInstanceOf(TerminalSession);
+          terminalView = workspaceElement.querySelector('terminal-view');
+        });
+
+        waitsForPromise(() => {
+          let terminalPromise = atom.commands.dispatch(terminalView, 'terminal:unfocus');
```

Comment:
I noticed that we can create the promise, and subsequently wait for it, in the same method. It's a little bit nicer to read in my opinion, but again not strongly opinionated.